File: cpp-sudoku/src/util/FileUtils.cpp
#include "FileUtils.hpp"
#include <fstream>
#include <sstream>
#include <iostream>

namespace just::demo::util {

std::string FileUtils::readFile(const std::filesystem::path& file) {
    std::ifstream ifs(file);
    if (!ifs.is_open()) {
        throw std::runtime_error("Cannot open file: " + file.string());
    }
    std::stringstream buffer;
    buffer << ifs.rdbuf();
    return buffer.str();
}

std::string FileUtils::readFile(const std::string& file) {
    return readFile(std::filesystem::path(file));
}

void FileUtils::writeFile(const std::filesystem::path& file, const std::string& data) {
    std::filesystem::create_directories(file.parent_path());
    std::ofstream ofs(file);
    if (!ofs.is_open()) {
        throw std::runtime_error("Cannot create file: " + file.string());
    }
    ofs << data;
}

void FileUtils::writeFile(const std::string& file, const std::string& data) {
    writeFile(std::filesystem::path(file), data);
}

void FileUtils::writeFile(const std::filesystem::path& file, const std::vector<uint8_t>& bytes) {
    std::filesystem::create_directories(file.parent_path());
    std::ofstream ofs(file, std::ios::binary);
    if (!ofs.is_open()) {
        throw std::runtime_error("Cannot create file: " + file.string());
    }
    ofs.write(reinterpret_cast<const char*>(bytes.data()), bytes.size());
}

void FileUtils::appendFile(const std::filesystem::path& file, const std::string& data) {
    std::filesystem::create_directories(file.parent_path());
    std::ofstream ofs(file, std::ios::app);
    if (!ofs.is_open()) {
        throw std::runtime_error("Cannot append to file: " + file.string());
    }
    ofs << data;
}

void FileUtils::appendFile(const std::string& file, const std::string& data) {
    appendFile(std::filesystem::path(file), data);
}

std::vector<std::filesystem::path> FileUtils::streamFiles(const std::filesystem::path& dir) {
    std::vector<std::filesystem::path> files;
    if (std::filesystem::is_directory(dir)) {
        for (const auto& entry : std::filesystem::recursive_directory_iterator(dir)) {
            if (std::filesystem::is_regular_file(entry)) {
                files.push_back(entry.path());
            }
        }
    }
    return files;
}

} // namespace just::demo::util

File: src/main/java/just/demo/util/FileUtils.java
package just.demo.util;

import static java.nio.charset.StandardCharsets.UTF_8;
import static java.nio.file.Files.createDirectories;
import static java.nio.file.Files.readString;
import static java.nio.file.Files.writeString;
import static java.nio.file.StandardOpenOption.APPEND;
import static java.nio.file.StandardOpenOption.CREATE;
import static java.util.Arrays.stream;

import java.io.File;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Stream;

import lombok.experimental.UtilityClass;

@UtilityClass
public class FileUtils {

  public static String readFile(File file) {
    return readFile(file.toPath());
  }

  public static String readFile(Path file) {
    try {
      return readString(file, UTF_8);
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }

  public static void writeFile(File file, String data) {
    writeFile(file.toPath(), data);
  }

  public static void writeFile(Path file, String data) {
    try {
      createDirectories(file.getParent());
      writeString(file, data, UTF_8);
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }

  public static void writeFile(Path file, byte[] bytes) {
    try {
      Files.write(file, bytes);
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }

  public static void appendFile(File file, String data) {
    try {
      createDirectories(file.toPath().getParent());
      writeString(file.toPath(), data, UTF_8, CREATE, APPEND);
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }

  public static Stream<File> streamFiles(File dir) {
    return dir.isDirectory() ? stream(dir.listFiles()).flatMap(FileUtils::streamFiles) : Stream.of(dir);
  }

  public static Stream<File> streamFiles(Path dir) {
    return streamFiles(dir.toFile());
  }
}
